trieste/utils/pareto.py

```diff
@@ -106,6 +113,164 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
+        # :return: True if the test point dominates or augments the Pareto front (boolean)
+
+        # if and only if the test point is at least in one dimension smaller
+        # for every point in the Pareto set
+        idx_dom_augm = tf.reduce_any(smaller, axis=1)
+        is_dom_augm = tf.reduce_all(idx_dom_augm)
+
+        return is_dom_augm
+
+    def divide_conquer_nd(self, front: TensorType, threshold: TensorType = 0) -> TensorType:
+        # Divide and conquer strategy to compute the cells covering the non-dominated region.
+        # Generic version: works for an arbitrary number of objectives.
+
+        len_front, number_of_objectives = front.shape
+        jitter = gpflow.config.default_jitter()
+        lower = tf.zeros([0, number_of_objectives], dtype=tf.int32)
+        upper = tf.zeros([0, number_of_objectives], dtype=tf.int32)
+
+        # Extend front with the ideal and anti-ideal point
+        min_pf = tf.reduce_min(front, axis=0, keepdims=True) - 1
+        max_pf = tf.reduce_max(front, axis=0, keepdims=True) + 1
+        pf_ext = tf.concat(
+            [min_pf, front, max_pf], axis=0
+        )  # Needed for early stopping check (threshold)
+
+        pf_ext_idx = tf.concat(
+            [
+                tf.zeros([1, number_of_objectives], dtype=tf.int32),
+                tf.argsort(front, axis=0) + 1,  # +1 as index zero is reserved for the ideal point
+                tf.ones([1, number_of_objectives], dtype=tf.int32) * len_front + 1,
+            ],
+            axis=0,
+        )
+
+        # Start with one cell covering the whole front
+        dc = tf.stack(
+            [
+                tf.zeros(number_of_objectives, dtype=tf.int32),
+                (int(pf_ext_idx.shape[0]) - 1) * tf.ones(number_of_objectives, dtype=tf.int32),
+            ],
+            axis=0,
+        )[None]
+
+        total_size = tf.reduce_prod(max_pf - min_pf)
+        # Start divide and conquer until we processed all cells
+
+        _, lower, upper = tf.while_loop(
+            lambda dc, lower, upper: dc.shape[0] > 0,
+            lambda dc, lower, upper: self._while_body(
+                dc,
+                lower,
+                upper,
+                number_of_objectives,
+                pf_ext,
+                pf_ext_idx,
+                jitter,
+                front,
+                total_size,
+                threshold,
+            ),
+            loop_vars=[dc, lower, upper],
+            shape_invariants=[
+                tf.TensorShape([None, 2, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+            ],
+        )
+        return BoundedVolumes(lower, upper)
+
+    def _while_body(
+        self,
+        dc,
+        lower,
+        upper,
+        number_of_objectives,
+        pf_ext,
+        pf_ext_idx,
+        jitter,
+        front,
+        total_size,
+        threshold,
+    ):
```

Comment:
if you make this function local to `divide_conquer_nd` it will keep the namespace of `Pareto` clearer. You'll also be able to use local variables like `number_of_objectives` without using a `lambda` which should make the code easier to read